File: runtime-src/Monk/Runtime/Pattern.hs
{-# LANGUAGE OverloadedStrings #-}

module Monk.Runtime.Pattern (patternParts, matches, globPaths, trimPattern, trimPatternParts, replaceLiteral) where

import Control.Exception (IOException, bracket, catch)
import Control.Monad (filterM)
import Data.Array (listArray, (!))
import Data.ByteString (ByteString)
import Data.ByteString qualified as B
import Data.IntSet qualified as S
import Data.List (sort)
import Data.Word (Word8)
import System.Posix.Directory.ByteString qualified as D
import System.Posix.Files.ByteString qualified as F

patternParts :: [ByteString] -> Either ByteString [(Bool, ByteString)]
patternParts [] = Right []
patternParts (active : value : rest) | active `elem` ["0", "1"] = ((active == "1", value) :) <$> patternParts rest
patternParts _ = Left "invalid pattern frames"

-- Backslashes produced by an unquoted expansion remain pattern syntax. Quote
-- protected metacharacters first, then tokenize, including across fragments.
data Token = Literal Word8 | Star | AnyByte | ByteClass Bool S.IntSet
  deriving (Eq)

tokens :: [(Bool, ByteString)] -> [Token]
tokens = tokenize False

tokenize :: Bool -> [(Bool, ByteString)] -> [Token]
tokenize pathname = parse . unescape . concatMap encode
  where
    encode (active, bytes) = concatMap (\c -> [92 | not active && c `elem` [33, 94, 45, 46, 58, 61, 42, 91, 93, 63, 92, 37, 35, 40, 124, 41]] <> [c]) (B.unpack bytes)
    unescape (92 : c : rest) = (False, c) : unescape rest
    unescape (c : rest) = (True, c) : unescape rest
    unescape [] = []
    parse [] = []
    parse ((True, 42) : rest) = Star : parse rest
    parse ((True, 63) : rest) = AnyByte : parse rest
    parse ((True, 91) : rest) = case bracketClass rest of
      Just (token, after)
        | not pathname || all ((/= 47) . snd) (take (length rest - length after) rest) -> token : parse after
        | otherwise -> Literal 91 : parse rest
      Nothing -> Literal 91 : parse rest
    parse ((_, c) : rest) = Literal c : parse rest

-- POSIX C-locale classes are byte classes; neither Unicode decoding nor the
-- host process locale participates in matching.
classBytes :: [Word8] -> S.IntSet
classBytes name = S.fromList [n | n <- [0 .. 255], selected n]
  where
    alpha n = n >= 65 && n <= 90 || n >= 97 && n <= 122
    digit n = n >= 48 && n <= 57
    selected n = case B.pack name of
      "alnum" -> alpha n || digit n
      "alpha" -> alpha n
      "ascii" -> n < 128
      "blank" -> n `elem` [9, 32]
      "cntrl" -> n < 32 || n == 127
      "digit" -> digit n
      "graph" -> n >= 33 && n <= 126
      "lower" -> n >= 97 && n <= 122
      "print" -> n >= 32 && n <= 126
      "punct" -> n >= 33 && n <= 126 && not (alpha n || digit n)
      "space" -> n `elem` [9, 10, 11, 12, 13, 32]
      "upper" -> n >= 65 && n <= 90
      "word" -> alpha n || digit n || n == 95
      "xdigit" -> digit n || n >= 65 && n <= 70 || n >= 97 && n <= 102
      _ -> False

bracketClass :: [(Bool, Word8)] -> Maybe (Token, [(Bool, Word8)])
bracketClass input = do
  let (negative, body) = case input of
        (True, c) : rest | c `elem` [33, 94] -> (True, rest)
        _ -> (False, input)
  (members, rest) <- collect True [] body
  let bytes = S.unions (ranges members)
  pure (if invalidRange members || S.member (-1) bytes then ByteClass False S.empty else ByteClass negative bytes, rest)
  where
    collect _ _ [] = Nothing
    collect False members ((True, 93) : rest) = Just (reverse members, rest)
    collect _ members remaining = do
      (member, rest) <- atom remaining
      collect False (member : members) rest
    atom ((True, 91) : (True, marker) : rest) | marker `elem` [58, 46, 61] =
      case special marker [] rest of
        Just (name, after) -> Just (Right (if marker == 58 then classBytes name else case name of [c] -> S.singleton (fromIntegral c); _ -> S.singleton (-1)), after)
        Nothing -> Just (Left (False, 91), (True, marker) : rest)
    atom (byte : rest) = Just (Left byte, rest)
    atom [] = Nothing
    special _ _ [] = Nothing
    special marker acc ((True, c) : (True, 93) : rest) | c == marker = Just (reverse acc, rest)
    special marker acc ((_, c) : rest) = special marker (c : acc) rest
    invalidRange (Left _ : Left (True, 45) : Right _ : _) = True
    invalidRange (_ : rest) = invalidRange rest
    invalidRange [] = False
    ranges (Left (_, low) : Left (True, 45) : Left (_, high) : rest) = S.fromList [fromIntegral low .. fromIntegral high] : ranges rest
    ranges (Left (_, c) : rest) = S.singleton (fromIntegral c) : ranges rest
    ranges (Right members : rest) = members : ranges rest
    ranges [] = []

matchTokens :: [Token] -> ByteString -> Bool
matchTokens ts input = S.member size (B.foldl' step (closure (S.singleton 0)) input)
  where
    size = length ts
    arr = listArray (0, size - 1) ts
    closure states = foldl (\s i -> if S.member i s && arr ! i == Star then S.insert (i + 1) s else s) states [0 .. size - 1]
    accepts AnyByte _ = True
    accepts (Literal a) b = a == b
    accepts (ByteClass negative members) c = S.member (fromIntegral c) members /= negative
    accepts Star _ = False
    step states c = closure (S.fromList (concatMap (\i -> if i == size then [] else if arr ! i == Star then [i] else [i + 1 | accepts (arr ! i) c]) (S.toList states)))

literal :: Token -> Bool
literal (Literal _) = True
literal _ = False

matches :: ByteString -> [(Bool, ByteString)] -> Bool
matches subject = (`matchTokens` subject) . tokens

globPaths :: [(Bool, ByteString)] -> IO [ByteString]
globPaths parts
  | all literal (tokenize True parts) = pure [B.concat (map snd parts)]
  | otherwise = do
      found <- walk "" (spelledComponents False (components (tokenize True parts)))
      pure $ if null found then [B.concat (map snd parts)] else sort found
  where
    components [] = [[]]
    components xs = let (a, b) = break (== Literal 47) xs in a : case b of [] -> []; _ : rest -> components rest
    -- Bash retains literal separators before its first expanded component,
    -- but joins the expanded suffix with one slash, including a trailing one.
    spelledComponents _ [] = []
    spelledComponents True ([] : rest@(_ : _)) = spelledComponents True rest
    spelledComponents expanded (part : rest) = part : spelledComponents (expanded || not (all literal part)) rest
    exists path = (F.getSymbolicLinkStatus path >> pure True) `catch` (\(_ :: IOException) -> pure False)
    directory path = (F.isDirectory <$> F.getFileStatus path) `catch` (\(_ :: IOException) -> pure False)
    listing path = bracket (D.openDirStream (if B.null path then "." else path)) D.closeDirStream (\stream -> let loop = do name <- D.readDirStream stream; if B.null name then pure [] else (name :) <$> loop in loop) `catch` (\(_ :: IOException) -> pure [])
    walk parent [] = do ok <- exists parent; pure [parent | ok]
    walk parent (component : rest)
      | null component = if null rest then do ok <- directory (if B.null parent then "." else parent); pure [parent | ok] else walk (parent <> "/") rest
      | all literal component = descend (parent <> B.pack [c | Literal c <- component]) rest
      | otherwise = do
          names <- listing parent
          let visible name = not (B.null name) && (B.head name /= 46 || take 1 component == [Literal 46]) && name `notElem` [".", ".."]
          chosen <- filterM (\name -> if null rest then pure True else directory (parent <> name)) (filter (\name -> visible name && matchTokens component name) names)
          concat <$> mapM (\name -> descend (parent <> name) rest) chosen
    descend path [] = walk path []
    descend path rest = walk (path <> "/") rest

-- | Byte parameter removal; the admitted pattern alphabet is literal bytes,
-- star, question mark and C-locale byte bracket expressions. Prefix/suffix selection follows Bash shortest/longest.
trimPattern :: Bool -> Bool -> ByteString -> ByteString -> ByteString
trimPattern prefix longest subject patternBytes = trimPatternParts prefix longest subject [(True, patternBytes)]

-- | Quoted fragments remain literal while adjacent active fragments retain
-- wildcard syntax, including bracket expressions crossing fragment boundaries.
trimPatternParts :: Bool -> Bool -> ByteString -> [(Bool, ByteString)] -> ByteString
trimPatternParts prefix longest subject parts =
  case filter matching candidates of
    n : _ -> if prefix then B.drop n subject else B.take (size - n) subject
    [] -> subject
  where
    size = B.length subject
    candidates = if longest then [size, size - 1 .. 0] else [0 .. size]
    compiled = tokens parts
    matching n = matchTokens compiled (if prefix then B.take n subject else B.drop (size - n) subject)

-- | Literal replacement never interprets replacement ampersands or backslashes.
replaceLiteral :: Bool -> ByteString -> ByteString -> ByteString -> ByteString
replaceLiteral allMatches subject needle replacement
  | B.null needle = subject
  | otherwise = case B.breakSubstring needle subject of
      (before, after)
        | B.null after -> before
        | otherwise -> before <> replacement <> (if allMatches then replaceLiteral True rest needle replacement else rest)
        where
          rest = B.drop (B.length needle) after

File: runtime/src/semantics/pattern.rs
//! C-locale byte patterns and pathname expansion. Protected fragments are
//! escaped before tokenization so brackets may span fragment boundaries.
use std::ffi::OsStr;
use std::os::unix::ffi::OsStrExt;
use std::path::Path;
pub type Parts = Vec<(bool, Vec<u8>)>;
pub fn pattern_parts(frames: &[Vec<u8>]) -> Result<Parts, Vec<u8>> {
    if !frames.len().is_multiple_of(2) {
        return Err(b"invalid pattern frames".to_vec());
    }
    frames
        .as_chunks::<2>()
        .0
        .iter()
        .map(|pair| match pair[0].as_slice() {
            b"0" => Ok((false, pair[1].clone())),
            b"1" => Ok((true, pair[1].clone())),
            _ => Err(b"invalid pattern frames".to_vec()),
        })
        .collect()
}
#[derive(Clone, PartialEq, Eq)]
enum Token {
    Literal(u8),
    Star,
    AnyByte,
    Class(bool, Box<[bool; 256]>),
}
enum Member {
    Byte(bool, u8),
    Set(Box<[bool; 256]>, bool),
}
fn class_bytes(name: &[u8]) -> Box<[bool; 256]> {
    let mut bytes = Box::new([false; 256]);
    for n in 0_u16..256 {
        let c = n as u8;
        let alpha = c.is_ascii_alphabetic();
        let digit = c.is_ascii_digit();
        bytes[usize::from(c)] = match name {
            b"alnum" => alpha || digit,
            b"alpha" => alpha,
            b"ascii" => c < 128,
            b"blank" => matches!(c, 9 | 32),
            b"cntrl" => c < 32 || c == 127,
            b"digit" => digit,
            b"graph" => (33..=126).contains(&c),
            b"lower" => c.is_ascii_lowercase(),
            b"print" => (32..=126).contains(&c),
            b"punct" => (33..=126).contains(&c) && !(alpha || digit),
            b"space" => matches!(c, 9 | 10 | 11 | 12 | 13 | 32),
            b"upper" => c.is_ascii_uppercase(),
            b"word" => alpha || digit || c == 95,
            b"xdigit" => c.is_ascii_hexdigit(),
            _ => false,
        };
    }
    bytes
}
fn bracket_class(input: &[(bool, u8)]) -> Option<(Token, usize)> {
    let negative = matches!(input.first(), Some((true, b'!' | b'^')));
    let mut i = usize::from(negative);
    let mut members = Vec::new();
    let mut closed = false;
    while i < input.len() {
        if !members.is_empty() && input[i] == (true, b']') {
            i += 1;
            closed = true;
            break;
        }
        if input[i] == (true, b'[')
            && input
                .get(i + 1)
                .is_some_and(|&(active, c)| active && matches!(c, b':' | b'.' | b'='))
        {
            let marker = input[i + 1].1;
            let mut end = i + 2;
            while end + 1 < input.len()
                && !(input[end] == (true, marker) && input[end + 1] == (true, b']'))
            {
                end += 1;
            }
            if end + 1 < input.len() {
                let name: Vec<u8> = input[i + 2..end].iter().map(|x| x.1).collect();
                let (set, invalid) = if marker == b':' {
                    (class_bytes(&name), false)
                } else {
                    let mut set = Box::new([false; 256]);
                    if name.len() == 1 {
                        set[usize::from(name[0])] = true;
                    }
                    (set, name.len() != 1)
                };
                members.push(Member::Set(set, invalid));
                i = end + 2;
                continue;
            }
            members.push(Member::Byte(false, b'['));
            i += 1;
            continue;
        }
        members.push(Member::Byte(input[i].0, input[i].1));
        i += 1;
    }
    if !closed {
        return None;
    }
    let invalid = members.windows(3).any(|w| {
        matches!(
            w,
            [Member::Byte(..), Member::Byte(true, b'-'), Member::Set(..)]
        )
    }) || members.iter().any(|m| matches!(m, Member::Set(_, true)));
    let mut bytes = Box::new([false; 256]);
    let mut at = 0;
    while at < members.len() {
        match &members[at..] {
            [
                Member::Byte(_, low),
                Member::Byte(true, b'-'),
                Member::Byte(_, high),
                ..,
            ] => {
                for c in *low..=*high {
                    bytes[usize::from(c)] = true;
                }
                at += 3;
            }
            [Member::Byte(_, c), ..] => {
                bytes[usize::from(*c)] = true;
                at += 1;
            }
            [Member::Set(set, _), ..] => {
                for c in 0..256 {
                    bytes[c] |= set[c];
                }
                at += 1;
            }
            [] => break,
        }
    }
    Some((
        if invalid {
            Token::Class(false, Box::new([false; 256]))
        } else {
            Token::Class(negative, bytes)
        },
        i,
    ))
}
fn tokenize(parts: &[(bool, Vec<u8>)], pathname: bool) -> Vec<Token> {
    let mut encoded = Vec::new();
    for (active, bytes) in parts {
        for &c in bytes {
            if !active && b"!^-.:=*[]?\\%#(|)".contains(&c) {
                encoded.push(b'\\');
            }
            encoded.push(c);
        }
    }
    let mut unescaped = Vec::new();
    let mut at = 0;
    while at < encoded.len() {
        let c = encoded[at];
        if c == b'\\' && at + 1 < encoded.len() {
            unescaped.push((false, encoded[at + 1]));
            at += 2;
        } else {
            unescaped.push((true, c));
            at += 1;
        }
    }
    let mut out = Vec::new();
    at = 0;
    while at < unescaped.len() {
        let (active, c) = unescaped[at];
        at += 1;
        match (active, c) {
            (true, b'*') => out.push(Token::Star),
            (true, b'?') => out.push(Token::AnyByte),
            (true, b'[') => {
                if let Some((token, used)) = bracket_class(&unescaped[at..])
                    && (!pathname || !unescaped[at..at + used].iter().any(|x| x.1 == b'/'))
                {
                    out.push(token);
                    at += used;
                    continue;
                }
                out.push(Token::Literal(b'['));
            }
            _ => out.push(Token::Literal(c)),
        }
    }
    out
}
fn closure(tokens: &[Token], states: &mut [bool]) {
    for i in 0..tokens.len() {
        if states[i] && tokens[i] == Token::Star {
            states[i + 1] = true;
        }
    }
}
struct Matcher<'a> {
    tokens: &'a [Token],
    states: Vec<bool>,
    next: Vec<bool>,
}
impl<'a> Matcher<'a> {
    fn new(tokens: &'a [Token]) -> Self {
        Self {
            tokens,
            states: vec![false; tokens.len() + 1],
            next: vec![false; tokens.len() + 1],
        }
    }
    fn matches(&mut self, subject: &[u8]) -> bool {
        let tokens = self.tokens;
        let states = &mut self.states;
        let next = &mut self.next;
        states.fill(false);
        states[0] = true;
        closure(tokens, states);
        for &c in subject {
            next.fill(false);
            for i in 0..tokens.len() {
                if !states[i] {
                    continue;
                }
                match &tokens[i] {
                    Token::Star => next[i] = true,
                    Token::AnyByte => next[i + 1] = true,
                    Token::Literal(b) => next[i + 1] |= *b == c,
                    Token::Class(negative, members) => {
                        next[i + 1] |= members[usize::from(c)] != *negative
                    }
                }
            }
            closure(tokens, next);
            std::mem::swap(states, next);
        }
        states[tokens.len()]
    }
}
pub fn matches(subject: &[u8], parts: &[(bool, Vec<u8>)]) -> bool {
    Matcher::new(&tokenize(parts, false)).matches(subject)
}
fn all_literal(tokens: &[Token]) -> bool {
    tokens.iter().all(|t| matches!(t, Token::Literal(_)))
}
fn path(bytes: &[u8]) -> &Path {
    Path::new(OsStr::from_bytes(bytes))
}
fn is_directory(bytes: &[u8]) -> bool {
    std::fs::metadata(path(bytes)).is_ok_and(|m| m.is_dir())
}
fn walk(parent: Vec<u8>, components: &[Vec<Token>], found: &mut Vec<Vec<u8>>) {
    let Some((component, rest)) = components.split_first() else {
        if std::fs::symlink_metadata(path(&parent)).is_ok() {
            found.push(parent);
        }
        return;
    };
    if component.is_empty() {
        if rest.is_empty() {
            if is_directory(if parent.is_empty() { b"." } else { &parent }) {
                found.push(parent);
            }
        } else {
            let mut parent = parent;
            parent.push(b'/');
            walk(parent, rest, found);
        }
        return;
    }
    if all_literal(component) {
        let mut joined = parent;
        joined.extend(component.iter().filter_map(|t| {
            if let Token::Literal(c) = t {
                Some(*c)
            } else {
                None
            }
        }));
        descend(joined, rest, found);
        return;
    }
    let Ok(entries) = std::fs::read_dir(path(if parent.is_empty() { b"." } else { &parent }))
    else {
        return;
    };
    let mut matcher = Matcher::new(component);
    for entry in entries.flatten() {
        let name = entry.file_name();
        let bytes = name.as_bytes();
        if bytes.is_empty()
            || bytes == b"."
            || bytes == b".."
            || (bytes[0] == b'.' && component.first() != Some(&Token::Literal(b'.')))
            || !matcher.matches(bytes)
        {
            continue;
        }
        let joined = [parent.as_slice(), bytes].concat();
        if rest.is_empty() || is_directory(&joined) {
            descend(joined, rest, found);
        }
    }
}
fn descend(mut path: Vec<u8>, rest: &[Vec<Token>], found: &mut Vec<Vec<u8>>) {
    if !rest.is_empty() {
        path.push(b'/');
    }
    walk(path, rest, found);
}
pub fn glob_paths(parts: &[(bool, Vec<u8>)]) -> Vec<Vec<u8>> {
    let spelling: Vec<u8> = parts
        .iter()
        .flat_map(|(_, bytes)| bytes.iter().copied())
        .collect();
    let tokens = tokenize(parts, true);
    if all_literal(&tokens) {
        return vec![spelling];
    }
    let raw: Vec<Vec<Token>> = tokens
        .split(|t| *t == Token::Literal(b'/'))
        .map(<[Token]>::to_vec)
        .collect();
    let mut expanded = false;
    let mut components = Vec::new();
    for (i, part) in raw.iter().enumerate() {
        if expanded && part.is_empty() && i + 1 < raw.len() {
            continue;
        }
        expanded |= !all_literal(part);
        components.push(part.clone());
    }
    let mut found = Vec::new();
    walk(Vec::new(), &components, &mut found);
    if found.is_empty() {
        vec![spelling]
    } else {
        found.sort();
        found
    }
}
pub fn trim_pattern_parts(
    prefix: bool,
    longest: bool,
    subject: &[u8],
    parts: &[(bool, Vec<u8>)],
) -> Vec<u8> {
    let tokens = tokenize(parts, false);
    let mut matcher = Matcher::new(&tokens);
    for i in 0..=subject.len() {
        let n = if longest { subject.len() - i } else { i };
        let candidate = if prefix {
            &subject[..n]
        } else {
            &subject[subject.len() - n..]
        };
        if matcher.matches(candidate) {
            return if prefix {
                subject[n..].to_vec()
            } else {
                subject[..subject.len() - n].to_vec()
            };
        }
    }
    subject.to_vec()
}
pub fn replace_literal(
    all_matches: bool,
    subject: &[u8],
    needle: &[u8],
    replacement: &[u8],
) -> Vec<u8> {
    if needle.is_empty() {
        return subject.to_vec();
    }
    let mut rest = subject;
    let mut out = Vec::new();
    while let Some(at) = rest.windows(needle.len()).position(|s| s == needle) {
        out.extend(&rest[..at]);
        out.extend(replacement);
        rest = &rest[at + needle.len()..];
        if !all_matches {
            break;
        }
    }
    out.extend(rest);
    out
}
